Brave Search: Enhance search result details with extra snippets

drobertson-dev opened the pull request:
**Description:** 

This update significantly improves the Brave Search Tool's utility within the LangChain library by enriching the search results it returns. The tool previously returned title, link, and snippet, with the snippet being a truncated 140-character description from the search engine. To make the search results more informative, this update enables extra_snippets by default and introduces additional result fields: title, link, description (enhancing and renaming the former snippet field), age, and snippets. The snippets field provides a list of strings summarizing the webpage, utilizing Brave's capability for more detailed search insights. This enhancement aims to make the search tool far more informative and beneficial for users.

**Issue:** N/A

**Dependencies:** No additional dependencies introduced.

**Twitter handle:** @davidalexr987

**Code Changes Summary:**

- Changed the default setting to include extra_snippets in search results.
- Renamed the snippet field to description to accurately reflect its content and included an age field for search results.
- Introduced a snippets field that lists webpage summaries, providing users with comprehensive search result insights.

**Backward Compatibility Note:**

The renaming of snippet to description improves the accuracy of the returned data field but may impact existing users who have developed integration's or analyses based on the snippet field. I believe this change is essential for clarity and utility, and it aligns better with the data provided by Brave Search.

**Additional Notes:**

This proposal focuses exclusively on the Brave Search package, without affecting other LangChain packages or introducing new dependencies.

drobertson-dev committed:
Enhance Brave Search Tool Results with Additional Information and Summaries

The current implementation of the Brave Search Tool returns a limited set of information (title, link, description) for each search result. The description field provides only a truncated text of 140 characters, which significantly limits the usefulness of the search results.

To make the Brave Search Tool more useful and informative, I propose the following changes:

- Change the default setting of extra_snippets to True.
- Modify the return structure of search results to include:
title: The title of the webpage.
link: The URL of the webpage.
description: The full description of the webpage (previously returned as snippet).
- age: The publication or last updated age of the webpage.
- snippets: A list of strings providing a detailed summary of the webpage.

This change is expected to significantly improve the utility and user experience of the Brave Search Tool.

hwchase17 reviewed (commented):
do all results have these arguments? what type is extra_snippets? to me this seems like a rather large change, so id almost rather have this be a feature flag (`verbose=True`or the like)

daver987 commented:
> do all results have these arguments? what type is extra_snippets? to me this seems like a rather large change, so id almost rather have this be a feature flag (`verbose=True`or the like)

Hi, yes all results have those arguments.  Extra snippets is what brave calls their summary of the web page, it's usually anywhere from 2-5 paragraphs worth of text.  They specifically added it for use in AI tools.  

In the current implementation the most you can ever get back is the title, link and 'snippet', which is just the description that you would see on the search results page.  

This was how I fixed it for myself, but your right a feature flag would be better.

ccurme commented on libs/community/langchain_community/utilities/brave_search.py:
```diff
@@ -48,8 +50,8 @@ def download_documents(self, query: str) -> List[Document]:
             Document(
-                page_content=item.get("description"),  # type: ignore[arg-type]
-                metadata={"title": item.get("title"), "link": item.get("url")},
+                page_content=item.get("extra_snippets"),  # type: ignore[arg-type]
```
is `extra_snippets` a list of strings? If so, is your proposal to collapse them into a single string for page_content?

ccurme commented on libs/community/langchain_community/utilities/brave_search.py:
```diff
@@ -48,8 +50,8 @@ def download_documents(self, query: str) -> List[Document]:
-                page_content=item.get("description"),  # type: ignore[arg-type]
-                metadata={"title": item.get("title"), "link": item.get("url")},
+                page_content=item.get("extra_snippets"),  # type: ignore[arg-type]
+                metadata={"title": item.get("title"), "link": item.get("url"), "description": item.get("description")},
```
Please consult our [contributing guide](https://python.langchain.com/docs/contributing/code) for instructions here (you may need to run `make format-diff`).

ccurme commented:
Closing as author is unresponsive, will re-open if needed.

daver987 commented:
@ccurme Sorry about that I thought it was rejected. 
Currently the search results only returns the description from the SERP, its like 2 sentences and the rest of the data gets stripped out.  I can look at it again and make a modification that isn't a breaking change and still allows users to access the extra data if they wish.

drobertson-dev committed:
Update the BraveSearchWrapper to include extra snippets in results

- Update 'run' and 'download_documents' methods to concatenate 'description' and 'extra_snippets' when available

- Improves data returned from Brave search API, with no breaking changes.

daver987 commented:
@ccurme  I’ve improved the brave search tool with very minimal changes.  

- The `run` and `download_documents` methods now concatenate description and extra_snippets when they're available.
This does not introduce any breaking changes but still improves the results when using the brave search api.
	
Please let me know if there are any further adjustments needed. Thanks!

ccurme reviewed (approved):
Thanks!

ccurme committed:
Merge branch 'master' into patch-1

ccurme committed:
format
